torch/distributed/optim/optimizer.py

```diff
@@ -137,10 +188,10 @@ def step(self, context_id):
         """
         dist_autograd._is_valid_context(context_id)
         rpc_futs = []
-        for optim in self.remote_optimizers:
+        for optimizer in self.remote_optimizers:
             rpc_futs.append(rpc.rpc_async(
-                optim.owner(),
+                optimizer.owner(),
                 _local_optimizer_step,
```

Comment:
Can we have two versions of `_local_optimizer_step`? If we're using a torchscript functional optimizer, we can have a torchscript `_local_optimizer_step` which means RPC will call an torchscript function and there would be no GIL anywhere. If it is an optimizer that is not torchscript we call a regular python `_local_optimizer_step`.

In the current version, we still end up holding GIL since `_local_optimizer_step` is not torchscript.

Reply:
hmm I see, this make sense, the problem of this is that `_LocalOptimizer` still contains `Lock()` objects to protect those who doesn't have a functional optimizer implementation. I could try adding a new `_LocalScriptOptimizer` to enable the use case you mentioned